### PR TITLE
add a nodes table

### PR DESCRIPTION
**Pull request by amdwit**
first step of adding nodes to the database

### CODE REVIEW

File: app/backend/src/couchers/models.py [hunk 1]
```diff
@@ -653,3 +653,19 @@ class InitiatedUpload(Base):
...
+    """
+    Node, i.e. geographical subdivision of the world
+    """
+
+    __tablename__ = "nodes"
+
+    id = Column(Integer, primary_key=True)
+
+    name = Column(String, nullable=False)
+    parent_node = Column(Integer, ForeignKey("nodes.id"))
```
Comment (aapeliv) [inline app/backend/src/couchers/models.py:668]: Should be nullable?

File: app/backend/src/couchers/models.py [hunk 2]
```diff
@@ -653,3 +653,19 @@ class InitiatedUpload(Base):
...
+    polygon = Column(Geometry(geometry_type="POLYGON", srid=4326), nullable=True)
```
Comment (aapeliv) [inline app/backend/src/couchers/models.py]: Shouldn't be nullable

File: app/pb/api.proto [hunk 3]
```diff
@@ -139,6 +139,11 @@ enum SmokingLocation {
   SMOKING_LOCATION_NO = 5;
 }
 
+message NullableInt32Value {
+  bool is_null = 1;
+  uint32 value = 2;
+}
+
```
Comment (aapeliv) [inline app/pb/api.proto]: These could be moved to a helper file if they start to be reused across files.

Also, this is now `uint32` even though it says `int32`!

File: app/pb/api.proto [hunk 4]
```diff
@@ -139,6 +139,11 @@ enum SmokingLocation {
   SMOKING_LOCATION_NO = 5;
 }
 
+message NullableInt32Value {
+  bool is_null = 1;
+  uint32 value = 2;
```
Comment (lucaslcode) [inline app/pb/api.proto]: This is the same as `NullableUInt32Value` atm

File: app/pb/communities.proto [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+message Point {
+  double lat = 1;
+  double lng = 2;
+}
+
+message Node {
+  int32 node_id = 1;
+  string name = 2;
+  org.couchers.api.core.NullableInt32Value parent_node = 3;
```
Comment (aapeliv) [inline app/pb/communities.proto]: I'm not sure if the this needs to be nullable here. We could use `0` (the default if no int32 is specified) to denote no parent.

Reason being that these (nullablesXXvalues) are a bit annoying to work with, and there should only be one node with no parent.

File: app/pb/communities.proto [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+import "pb/api.proto";
+
+message Point {
+  double lat = 1;
+  double lng = 2;
+}
+
+message Node {
+  int32 node_id = 1;
```
Comment (lucaslcode) [inline app/pb/communities.proto:13]: All other IDs are uint64s. I remember you said this shouldn't be the case, but maybe changing it should be it's own PR.

Comment (aapeliv) [inline app/pb/communities.proto:13]: Other proto files use `uint64` for ids (and `uint32` for counts, by the way). This is poorly documented, sorry!

File: app/pb/communities.proto (hunk 5 above)
Comment (lucaslcode) [inline app/pb/communities.proto]: I don't think this needs to be nullable, no parent could just be 0 since autoincrementing IDs start from 1 (right?)

File: app/pb/communities.proto [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  string name = 2;
+  org.couchers.api.core.NullableInt32Value parent_node = 3;
+  repeated Point polygon = 4;
```
Comment (lucaslcode) [inline app/pb/communities.proto:16]: Is this the center of the node?

File: app/pb/communities.proto [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+package org.couchers.api.communities;
+
+import "google/protobuf/timestamp.proto";
+
+import "pb/api.proto";
+
+message Point {
+  double lat = 1;
+  double lng = 2;
+}
```
Comment (aapeliv) [inline app/pb/communities.proto:10]: This is OK. At some point we will need some more message types for geo data (in a separate file, presumably). There are a few around that we should have a look at.

E.g. `message Polygon { repeated Point = 1; }` would be good.

**Review by aapeliv — COMMENTED** [review]
Thanks, this is a good start.

I think it would be good to have these bundled up in a larger PR though, so we can see how it all works together. Otherwise you might end up making a whole bunch of changes to the tables, which will make a lot of migrations, as well as make it hard to make modifications and judge the full feature as a whole.

**Review by lucaslcode — COMMENTED** [review]
Thanks for starting this! I wonder if it might be better to start with the API we definitely need rather than basing that on the table structure. In this case, to me it's not yet clear how much of 'node' actually needs to be in the API. It could end up being rolled into clusters (I guess that's the new name of groups?) or pages.

**Comment by lucaslcode** [comment]
Ooops looks like we were both looking :D

File: app/pb/communities.proto (hunk 7 above)
Comment (amdwit) [reply app/pb/communities.proto:16]: It is a list of points. Or do I understand it wrong? This means there is a property `polygon` which is a list of `Points`.

File: app/backend/src/couchers/models.py (hunk 1 above)
Comment (amdwit) [reply app/backend/src/couchers/models.py:668]: The top (world-wide) node does not have a parent. An alternative is to have it point to itself. Or we use zero. Ok, zero it will be.

File: app/pb/api.proto (hunk 3 above)
Comment (amdwit) [reply app/pb/api.proto]: solved

File: app/pb/communities.proto (hunk 6 above)
Comment (amdwit) [reply app/pb/communities.proto:13]: What should be used then? In models.py is this id described as Integer which translates to PostgreSQL as an int32. So it is best to use that here too. Why using uint32 or uint64 instead?

**Comment by amdwit** [comment]
I will make one new PR with all the communities tables in models.py and a proto for communities. I will coordinate with Christians work on threads.

File: app/pb/communities.proto (hunk 6 above)
Comment (amdwit) [reply app/pb/communities.proto:13]: PostgreSQL does not have unsigned integers. And in models.py is Integer used. That's why I think int32 is best.

**Comment by amdwit** [comment]
Is there anything known about what the api should look like?

File: app/pb/communities.proto (hunk 6 above)
Comment (kalvdans) [reply app/pb/communities.proto:13]: The largest dataset at https://simplemaps.com/data/world-cities contains 4.2 million places, so 31 bits should be enough to identify a unique node.

File: app/backend/src/couchers/models.py (hunk 1 above)
Comment (aapeliv) [reply app/backend/src/couchers/models.py:668]: Db calls might be easier with NULL for the world node

File: app/pb/communities.proto (hunk 6 above)
Comment (aapeliv) [reply app/pb/communities.proto:13]: We should stick to consistency and not pick id sizes based on the max number of entries. A few bytes here or there don't matter in this context.

Comment (amdwit) [reply app/pb/communities.proto:13]: Does this mean you want the datatype in models.py to be changed to `BigInteger`? Integer is 32 bit and BigInteger is 64 bit. Or another question: shouldn't the datatypes in model.py match those in the proto files?

Comment (aapeliv) [reply app/pb/communities.proto:13]: If someone would like to do the refactoring job of replacing all the integer `id`s in the models with big serials, that's OK. But it doesn't matter if it's the same as elsewhere and uses `Integer` for now.

Comment (amdwit) [reply app/pb/communities.proto:13]: Ok, so I understand you would like to use uint64 here, which I find strange as the model and db uses int32 and in js it is translated as float64 which does not cover uint64. To keep it consistently wrong is not a big enough reason for me to not do it right from now on. The future refactoring job will grow and grow. It will also not surprise me if newer grpc versions will start translating uint64 as bigint in js.

Comment (aapeliv) [reply app/pb/communities.proto:13]: Then please refactor all `id`s in `models.py` to be big serials and we'll start doing it right everywhere. That's better as a separate PR off `develop`, I'd say

File: app/pb/communities.proto (hunk 8 above)
Comment (amdwit) [reply app/pb/communities.proto:10]: Don't we need a field name in there?

Comment (aapeliv) [reply app/pb/communities.proto:10]: Yeah we do :P. My bad.

File: app/pb/communities.proto (hunk 7 above)
Comment (lucaslcode) [reply app/pb/communities.proto:16]: My bad, I missed `repeated`

**Comment by kalvdans** [comment]
> Is there anything known about what the api should look like?

I'll propose this just to get the discussion going:
```
// Returns immediate children of the given node. passing node_id 0 means get all top-level nodes.
rpc GetGeographicalNodeChildren(uint64 node_id) returns (repeated Node);

// Create a new node. Throws PERMISSION_DENIED if you are not an administrator of the group associated with parent node.
rpc CreateGeographicalNode(Node node) returns (uint64 node_id);

// Delete a node. Throws PERMISSION_DENIED if you are not an administrator of the parent node. Throws FAILED_PRECONDITION if there are still groups attached to this node.
rpc DeleteGeographicalNode(uint64 node_id) returns (google.protobuf.Empty);
```

**Comment by amdwit** [comment]
Enough for now.
